### PR TITLE
Filter graph

### PR DESCRIPTION
**Pull request by alangerak**
Added filtering for the graph.(https://github.com/ProgrammingLife5/LifeTiles/issues/53)

Currently there is code to draw Edges but this at the moment broken.

Also, there are tests added for the classes inside the graph.view package (https://github.com/ProgrammingLife5/LifeTiles/issues/98) 
And some styling code, (expect for the Font) is moved to a stylesheet (https://github.com/ProgrammingLife5/LifeTiles/issues/60)

### CODE REVIEW

File: src/main/java/nl/tudelft/lifetiles/graph/model/GraphContainer.java [hunk 1]
```diff
@@ -81,12 +103,38 @@ private void alignGraph() {
...
+     *
+     * @param visibleSequences
+     *            the sequences to display
+     */
+    public final void setVisible(final Set<Sequence> visibleSequences) {
+
+        // Find out which vertices are visible now
+        Set<SequenceSegment> vertices = new TreeSet<SequenceSegment>();
+        for (Sequence seq : visibleSequences) {
+            Iterator<SequenceSegment> itSegments = seq.getSegments().iterator();
```
Comment (jorenham) [inline src/main/java/nl/tudelft/lifetiles/graph/model/GraphContainer.java]: An enhanced for-loop would be cleaner than an iterator.

File: src/main/java/nl/tudelft/lifetiles/graph/model/GraphContainer.java [hunk 2]
```diff
@@ -97,7 +145,35 @@ private void findMutations(final Sequence reference) {
...
+        return copy;
+    }
+
+    /**
+     * Get if the graph in this container is recently changed.
+     *
+     * @return true if recently changed, else false
+     */
+    public final boolean isChanged() {
+        if (changed) {
```
Comment (jorenham) [inline src/main/java/nl/tudelft/lifetiles/graph/model/GraphContainer.java]: Why are you changing changed to false in this getter?

File: src/main/java/nl/tudelft/lifetiles/graph/view/EdgeLine.java [hunk 3]
```diff
@@ -208,24 +136,28 @@ public final Line getLine() {
...
      * @param destination
      *            the right bound
-     * @return true on y coordinate is outside of the range
+     * @return true if source lies indeed lies above or below
      */
     private Boolean outofboundsY(final Bounds source, final Bounds destination) {
-        if (source.getMinY() > destination.getMinY() && source.getMinY() > destination.getMaxY()
-                && source.getMaxY() > destination.getMaxY()
-                && source.getMaxY() > destination.getMinY()) {
+        // Check if top of source is below the destination
```
Comment (jorenham) [inline src/main/java/nl/tudelft/lifetiles/graph/view/EdgeLine.java]: You can merge the `if`'s by using `!=`.

File: src/main/java/nl/tudelft/lifetiles/graph/model/BucketCache.java [hunk 4]
```diff
@@ -68,7 +68,7 @@ public BucketCache(final int numberBuckets,
      */
     private void cacheGraph() {
         Timer timer = Timer.getAndStart();
-        
+
```
Comment (jorenham) [inline src/main/java/nl/tudelft/lifetiles/graph/model/BucketCache.java]: Useless diff

File: src/main/java/nl/tudelft/lifetiles/graph/view/EdgeLine.java [hunk 5]
```diff
@@ -47,89 +51,48 @@ public EdgeLine(final VertexView source, final VertexView destination) {
...
-        if (denom == 0 || (snumer < 0) == denomPositive
-                || (tnumer < 0) == denomPositive
-                || (snumer > denom) == denomPositive
-                || (tnumer > denom) == denomPositive) {
-            return -1;
-        }
+    private void drawCrossLine(final Bounds boundFrom, final Bounds boundTo) {
+
+        double fromX = boundFrom.getMaxX();
+        double fromY = boundFrom.getMinY() + (boundFrom.getHeight() / 2);
```
Comment (jorenham) [inline src/main/java/nl/tudelft/lifetiles/graph/view/EdgeLine.java]: Double code, `boundFrom.getHeight() / 2` needs a variable

File: src/main/java/nl/tudelft/lifetiles/graph/view/EdgeLine.java [hunk 6]
```diff
@@ -146,56 +109,21 @@ private void drawCrossLine(final VertexView source,
      */
     private void drawLine(final double startX, final double startY,
             final double endX, final double endY) {
+
+        Circle head = new Circle(HEAD_RADIUS, Color.RED);
```
Comment (jorenham) [inline src/main/java/nl/tudelft/lifetiles/graph/view/EdgeLine.java]: The colors should be constants.

File: src/main/java/nl/tudelft/lifetiles/graph/view/EdgeLine.java [hunk 7]
```diff
@@ -47,89 +51,48 @@ public EdgeLine(final VertexView source, final VertexView destination) {
...
+                || boundFrom.getMinY() > boundTo.getMinY()) {
 
-        // Check if the head has to be moved slightly up and down
-        // to display it more clearly
-        if (line.getEndY() - line.getStartY() > 0) {
-            head.setCenterY(line.getEndY() - head.getRadius());
-        }
-        if (line.getEndY() - line.getStartY() < 0) {
-            head.setCenterY(line.getEndY() + head.getRadius());
+            startY = boundFrom.getMinY() + boundFrom.getHeight() / 2;
```
Comment (jorenham) [inline src/main/java/nl/tudelft/lifetiles/graph/view/EdgeLine.java]: `boundFrom.getHeight() / 2` could use a variable

File: src/main/java/nl/tudelft/lifetiles/graph/view/EdgeLine.java [hunk 8]
```diff
@@ -47,89 +51,48 @@ public EdgeLine(final VertexView source, final VertexView destination) {
...
-        double toX = destination.getLayoutX();
-        double toY = destination.getLayoutY() + (boundTo.getHeight() / 2);
+    private void drawStraightLine(final Bounds boundFrom, final Bounds boundTo) {
 
-        drawLine(fromX, fromY, toX, toY);
+        double startY = 0;
 
-        head.setCenterX(line.getEndX() - head.getRadius());
+        // Decide from which bound the Y value should be taken from
+        if (boundFrom.getMaxY() < boundTo.getMaxY()
```
Comment (jorenham) [inline src/main/java/nl/tudelft/lifetiles/graph/view/EdgeLine.java]: Rewrite using `!=`

File: src/test/java/nl/tudelft/lifetiles/graph/model/FactoryProducerTest.java [hunk 9]
```diff
@@ -13,7 +14,9 @@
 
     @Test
     public void testGetWithParam() {
-        GraphFactory<String> gf = FactoryProducer.getFactory("JGraphT");
+        GraphFactory<SequenceSegment> gf = FactoryProducer
```
Comment (jorenham) [inline src/test/java/nl/tudelft/lifetiles/graph/model/FactoryProducerTest.java:17]: Should use overload method `getFactory()` instead

File: src/main/java/nl/tudelft/lifetiles/graph/model/BucketCache.java (hunk 4 above)
Comment (albert-smit) [inline src/main/java/nl/tudelft/lifetiles/graph/model/BucketCache.java]: no, it fixes a checkstyle error that was left in the code

File: src/main/resources/styles/SequenceStyle.css [hunk 10]
```diff
@@ -6,7 +6,7 @@
     -fx-border-style: solid;
     -fx-background-color: #FFFFFF;
 }
-
+
```
Comment (albert-smit) [inline src/main/resources/styles/SequenceStyle.css:9]: this is a useless diff

**Comment by albert-smit** [comment]
looks good when the above issues are fixed

**Comment by rutgervandenberg** [comment]
- Graph.java
  - Creating a getInternalGraph in the interface is a very bad idea. It creates dependencies on implementations. For example, if we merge this we can never remove the JGraphT library. Even worse, we now need to include it in every potential implementation of Graph. You should instead add it to the particular implemention and make it private. Thats where you want to use it anyways.
  - deepCopy might be slightly confusing. Since you're providing a GraphFactory the resulting graph could easily use a different graph library than the original. Further, ou're now using a GraphFactory of a (potentially) different type in the implementation. You should move this to the factory so that it is clear that you're creating an instance of that particular implementation. 
- GraphContainer:
  - getSegments: should rename this to getVisibleSegments or something like that, since it only returns currently visible segments. 
- JGraphTGraphAdapter: 
  - The new constructor should be (re)moved. 
    - First off, this depends on the provided Graph being of the JGraphT type, which you can't guarantee. 
    - Second, this breaks the contract of Graph. getSources and getSinks return empty sets, even if they should not. Further, addEdge(int, int) no longer works since vertexIdentifiers is empty. 
- SequenceSegment: 
  - copySources/copySequence: Why not just use copy constructors instead of your own method for this?

File: src/main/java/nl/tudelft/lifetiles/graph/model/jgrapht/JGraphTGraphAdapter.java [hunk 11]
```diff
@@ -252,6 +293,41 @@ private void removeEdge(final Edge<V> edge) {
...
+     */
+    @Override
+    public final Graph<V> deepcopy(final GraphFactory<V> gfact) {
+        Graph<V> copygraph = gfact.getGraph();
+
+        Map<Object, Object> convertVertices = new HashMap<Object, Object>();
+
+        for (V vertex : this.getAllVertices()) {
+            try {
+                Method method;
```
Comment (jorenham) [inline src/main/java/nl/tudelft/lifetiles/graph/model/jgrapht/JGraphTGraphAdapter.java]: Redundant line

File: src/main/java/nl/tudelft/lifetiles/graph/view/TileView.java [hunk 12]
```diff
@@ -127,12 +162,33 @@ private Color sequenceColor(final Mutation mutation) {
      * @param color
      *            the colour of the vertex
      */
-    private void drawVertex(final String text, final double xcoord,
+
+    private void drawVertex(final SequenceSegment segment, final double xcoord,
```
Comment (jorenham) [inline src/main/java/nl/tudelft/lifetiles/graph/view/TileView.java]: Too many arguments, `xcoord` and `ycoord` could be combined, same goes for `width` and `height`

**Comment by jorenham** [comment]
Looks good now

**Pull request merged**